### PR TITLE
executor: mutex snapshot with txn

### PR DESCRIPTION
**Pull request by xhebox**
Signed-off-by: xhe <xw897002528@gmail.com>

### What problem does this PR solve?

Problem Summary: Should not be able to set `tidb_snapshot` and `txn_read_ts` in the txn. And `start txn as of` should set `tidb_snapshot` into 0 as `set txn as of`.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- Forbid history read and changing stale transaction characteristics in the transaction. And stale transaction will reset tidb_snapshot into 0.

### CODE REVIEW

File: executor/stale_txn_test.go [hunk 1]
```diff
@@ -291,31 +292,54 @@ func (s *testStaleTxnSuite) TestStalenessAndHistoryRead(c *C) {
...
 	tk.MustExec(`set @@tidb_snapshot="2020-10-08 16:45:26";`)
 	c.Assert(tk.Se.GetSessionVars().SnapshotTS, Equals, uint64(419993151340544000))
+	c.Assert(tk.Se.GetSessionVars().SnapshotInfoschema, NotNil)
 	c.Assert(tk.Se.GetSessionVars().TxnReadTS.PeakTxnReadTS(), Equals, uint64(0))
+
+	c.Assert(tk.Se.GetSessionVars().SetSystemVar(variable.TiDBTxnReadTS, ""), IsNil)
+
+	// test start txn will flush/mutex tidb_snapshot
+	tk.MustExec(`set @@tidb_snapshot="2020-10-08 16:45:26";`)
+	c.Assert(tk.Se.GetSessionVars().SnapshotTS, Equals, uint64(419993151340544000))
```
Comment (Yisaer) [inline executor/stale_txn_test.go:316]: I think we can set `tidb_snapshot`/ `set transaction read only as of timestamp` / `start transaction read only as of timestamp` with different timestamp ( ddl operation between each timestamp) and assert the SchemaVersion of SnapshotInfoSchema.

**Review by Yisaer — COMMENTED** [review]
Rest LGTM

File: executor/stale_txn_test.go (hunk 1 above)
Comment (xhebox) [reply executor/stale_txn_test.go]: What about now?

**Review by Yisaer — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**